src/mavsdk/core/mavsdk_impl.cpp

```diff
@@ -612,30 +615,31 @@ void MavsdkImpl::make_system_with_component(
 void MavsdkImpl::notify_on_discover()
 {
     std::lock_guard<std::recursive_mutex> lock(_systems_mutex);
-    if (_new_system_callback) {
-        auto temp_callback = _new_system_callback;
-        call_user_callback([temp_callback]() { temp_callback(); });
-    }
+    _new_system_callbacks.queue([this](const auto& func) { call_user_callback(func); });
 }
 
 void MavsdkImpl::notify_on_timeout()
 {
     std::lock_guard<std::recursive_mutex> lock(_systems_mutex);
-    if (_new_system_callback) {
-        auto temp_callback = _new_system_callback;
-        call_user_callback([temp_callback]() { temp_callback(); });
-    }
+    _new_system_callbacks.queue([this](const auto& func) { call_user_callback(func); });
 }
 
-void MavsdkImpl::subscribe_on_new_system(const Mavsdk::NewSystemCallback& callback)
+Mavsdk::NewSystemHandle
+MavsdkImpl::subscribe_on_new_system(const Mavsdk::NewSystemCallback& callback)
 {
     std::lock_guard<std::recursive_mutex> lock(_systems_mutex);
-    _new_system_callback = callback;
+    const bool is_first_subscription = _new_system_callbacks.empty();
+    const auto handle = _new_system_callbacks.subscribe(callback);
 
-    if (_new_system_callback != nullptr && is_any_system_connected()) {
-        auto temp_callback = _new_system_callback;
-        call_user_callback([temp_callback]() { temp_callback(); });
+    if (is_first_subscription && is_any_system_connected()) {
```

Comment:
Hmm :thinking:.

So before we would set the `_new_system_callback`, and then we would check if it's not null.

Now we check if it's empty before setting it. This means we only notify of existing systems when adding the first callback but not on subsequent ones.

I believe I had some reasoning for that but maybe not, and it should really just be `callback != nullptr`, so on each new callback registered we call it for any existing system.

Reply:
I think it was to immediately call the callback if a system was already connected and then someone subscribed to the subscribe_on_new_system(), which makes sense because otherwise it would wait until the state of the system connection changed before it'd call the callback.

So every new subscriber's callback gets queued, and additionally when subscribing the callback should be immediately called(on the new subscriber's callback) if a system was already connected when subscribing.

Reply:
Should we still not check if the callback is nullptr? What is the new expected behavior when a subscriber tries to subscribe with a nullptr?
 - Do we just ignore the callback
 - Do we clear out the entire callback list (also the callbacks of every other subscriber)
 - Do we want to clear out all the callbacks added by that particular subscriber (we would need to map the callbacks to the subscriber)

Reply:
> What is the new expected behavior when a subscriber tries to subscribe with a nullptr?

I would just ignore it. The other two are not intuitive to me :sweat_smile:.

Reply:
I would clear out the entire list with nullptr, and issue a warning, but that's already done in CallbackList internally, see:
https://github.com/mavlink/MAVSDK/pull/1768/files#diff-16f795ac2a0d8f3ef2a46e2b3ded55f6d3f62d0e3e3be6b01581e37f0d38d744R80

However, the tricky point is this as @devbharat writes, and I didn't realize that, so thanks!

> the callback should be immediately called(only the new subscriber's callback) if a system was already connected when subscribing.

So, I need to add/implement that logic.

Reply:
Done in [debb0bf](https://github.com/mavlink/MAVSDK/pull/1768/commits/debb0bfd70721643b0e8e2c2630510cc1a58c274).